subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultMavenArtifactRepository.java

```diff
@@ -354,4 +370,29 @@ public void artifact() {
         }
     }
 
+    private static class SnapshotsOnly implements Action<ArtifactResolutionDetails> {
+        public final static SnapshotsOnly INSTANCE = new SnapshotsOnly();
+        @Override
+        public void execute(ArtifactResolutionDetails artifactResolutionDetails) {
+            if (!artifactResolutionDetails.isVersionListing()) {
+                ModuleComponentIdentifier componentId = artifactResolutionDetails.getComponentId();
+                if (!componentId.getVersion().endsWith("-SNAPSHOT")) {
+                    artifactResolutionDetails.notFound();
+                }
+            }
+        }
+    }
+
+    private static class ReleasesOnly implements Action<ArtifactResolutionDetails> {
+        public final static ReleasesOnly INSTANCE = new ReleasesOnly();
+        @Override
+        public void execute(ArtifactResolutionDetails artifactResolutionDetails) {
+            if (!artifactResolutionDetails.isVersionListing()) {
+                ModuleComponentIdentifier componentId = artifactResolutionDetails.getComponentId();
+                if (componentId.getVersion().endsWith("-SNAPSHOT")) {
+                    artifactResolutionDetails.notFound();
+                }
+            }
+        }
```

Comment:
This isn't going to exclude `SNAPSHOT` artifacts from being resolved via a dynamic version. So if I use `org:foo:+` I may still get a SNAPSHOT module being resolved from a repository with `RELEASES_ONLY`.

I recommend removing these Maven convenience methods from the PR, and adding convenience methods in a later story.

Reply:
I don't understand your comment here, I'm probably missing something. We're working at the repository access level, there are only 2 operations here:

1. give me the list of versions for module "org:foo". It's independent of having snapshots or not. It says "is it possible to find module `org:foo` in this repository?". Yes, it is, so it's going to return a version list, which, yes, may contain snapshots, but it's irrelevant.
2. tell me if version "X" is available in the repository. This is what is implemented here: if the version is `-SNAPSHOT`, then we will fail resolving, so there will not be any component metadata for this snapshot version. If this is the case, how would you resolve artifacts?

Reply:
I think we should avoid exposing these low-level implementation details in the filter API.

But my point is that if a user specifies `RELEASES_ONLY` and have a dependency on `org:foo:+`, then they may get `org:foo:1.0-SNAPSHOT` resolved from the repository. If the dependency is on `org:foo:1.0-SNAPSHOT` then they won't get it resolved from the repository. This inconsistency isn't good.

Reply:
I don't think this is true. If you write:

`conf "org:foo:1.+"`

then, what we do is that we list the versions for the `RELEASES_ONLY` server, which, yes, _may_ include snapshots (but in practice it would not). Even if it does, when `resolveComponentMetaData` is called, we check the version, which happens to contain `-SNAPSHOT`. It's a version that we get there, not a selector. So it's skipped. I think this would be true if I used a selector like what you suggest, but I don't. I'm working at the repository access level, so what we get there is resolved versions.

Reply:
(have a test showing this works as expected)